index/encoding/doc/doc.go

```diff
@@ -0,0 +1,274 @@
+// furnished to do so, subject to the following conditions:
+//
+// The above copyright notice and this permission notice shall be included in
+// all copies or substantial portions of the Softwarw.
+//
+// THE SOFTWARE IS PROVIDED "AS IS", WITHOUT WARRANTY OF ANY KIND, EXPRESS OR
+// IMPLIED, INCLUDING BUT NOT LIMITED TO THE WARRANTIES OF MERCHANTABILITY,
+// FITNESS FOR A PARTICULAR PURPOSE AND NONINFRINGEMENT. IN NO EVENT SHALL THE
+// AUTHORS OR COPYRIGHT HOLDERS BE LIABLE FOR ANY CLAIM, DAMAGES OR OTHER
+// LIABILITY, WHETHER IN AN ACTION OF CONTRACT, TORT OR OTHERWISE, ARISING FROM,
+// OUT OF OR IN CONNECTION WITH THE SOFTWARE OR THE USE OR OTHER DEALINGS IN
+// THE SOFTWARE.
+
+package doc
+
+import (
+	"errors"
+	"fmt"
+	"io"
+
+	"github.com/m3db/m3ninx/doc"
+	"github.com/m3db/m3ninx/index/encoding"
+)
+
+// Documents file format metadata.
+const (
+	magicNumber = 0x6D33D0C5
+
+	fileFormatVersion = 1
+
+	headerSize = 0 +
+		4 + // magic number
+		4 + // version
+		0
+
+	trailerSize = 0 +
+		8 + // number of documents
+		4 + // checksum of payload
+		0
+)
+
+const initialEncoderSize = 1024
+
+var (
+	errEmptyDocument      = errors.New("cannot write an empty document")
+	errInvalidMagicNumber = errors.New("encountered invalid magic number")
+	errChecksumMismatch   = errors.New("encountered invalid checksum")
+)
+
+type writer struct {
+	version uint32
+
+	writer io.Writer
+	enc    *encoding.Encoder
+
+	docs     uint64
+	checksum *encoding.Checksum
+}
+
+// NewWriter returns a new Writer for serializing documents to an underlying io.Writer.
+func NewWriter(w io.Writer) Writer {
+	return &writer{
+		version:  fileFormatVersion,
+		writer:   w,
+		enc:      encoding.NewEncoder(initialEncoderSize),
+		checksum: encoding.NewChecksum(),
+	}
+}
+
+func (w *writer) Open() error {
+	w.enc.PutUint32(magicNumber)
+	w.enc.PutUint32(w.version)
+	if err := w.write(); err != nil {
+		return err
+	}
+	return nil
+}
+
+func (w *writer) Write(d doc.Document) error {
+	if len(d.Fields) == 0 {
+		return errEmptyDocument
+	}
+
+	w.enc.PutUvarint(uint64(len(d.Fields)))
+
+	for _, f := range d.Fields {
+		w.enc.PutBytes(f.Name)
+		w.enc.PutBytes(f.Value)
+	}
+
+	if err := w.write(); err != nil {
+		return err
+	}
+
+	w.docs++
+	return nil
+}
+
+func (w *writer) Close() error {
+	w.enc.PutUint64(w.docs)
+	if err := w.write(); err != nil {
+		return err
+	}
+
+	w.enc.PutUint32(w.checksum.Get())
+	if err := w.write(); err != nil {
+		return err
+	}
+
+	return nil
+}
+
+func (w *writer) write() error {
+	b := w.enc.Bytes()
+	n, err := w.writer.Write(b)
+	if err != nil {
+		return err
+	}
+	if n < len(b) {
+		return io.ErrShortWrite
+	}
+	w.checksum.Update(b)
+	w.enc.Reset()
+	return nil
+}
+
+type reader struct {
+	version uint32
+
+	data []byte
+	dec  *encoding.Decoder
+
+	curr     uint64
+	total    uint64
+	checksum *encoding.Checksum
+}
+
+// NewReader returns a new Reader for deserializing documents from a byte slice.
+func NewReader(data []byte) Reader {
```

Comment:
This looks like you need to read the entire dataset into memory before iterating on it. Anyway to stream the data as you read it?